### PR TITLE
ICMSLST-1917 Cover Letter PDF

### PR DESCRIPTION
**Pull request by mikeylockwood**

### CODE REVIEW

File: web/templates/pdf/import/cover-letter.html [hunk 1]
```diff
@@ -0,0 +1,83 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <link rel="icon" href="{{ static('web/img/favicon.png') }}" type="image/png">
```
Comment (MattHolmes123) [inline web/templates/pdf/import/cover-letter.html:4]: Do we need this?

File: web/templates/pdf/import/cover-letter.html [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    <link rel="stylesheet" href="{{ static('web/css/pdfs/cover-letter.css') }}"/>
+    {% if preview %}
+      <link rel="stylesheet" href="{{ static('web/css/pdfs/preview.css') }}"/>
+    {% endif %}
+    {%- if page_title %}
+     {% set title=page_title %}
+    {% else %}
+      {% set title='Import Case Management System' %}
+    {% endif -%}
```
Comment (MattHolmes123) [inline web/templates/pdf/import/cover-letter.html:13]: This if else block feels like you've copied it from somewhere else.
Do we need a title for a cover letter template?

File: web/templates/pdf/import/cover-letter.html [hunk 3]
```diff
@@ -0,0 +1,83 @@
+<!DOCTYPE html>
+<html>
```
Comment (MattHolmes123) [inline web/templates/pdf/import/cover-letter.html:2]: Missing required 'lang' attribute 
`<html lang="en">`

File: web/templates/pdf/import/cover-letter.html [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        alt="DBT Logo"
+        class="header-block"
+        src="{{ static('web/img/dbt_red_transparent_bg.png') }}" />
+
+        <div id="address-text" class="header-block">
+          <p class="address-line">Import Licencing Branch</p>
+          <p class="address-line">Queensway House</p>
+          <p class="address-line">West Precinct</p>
+          <p class="address-line">Billingham</p>
+          <p class="address-line">TS23 2NF</p>  <!-- /PS-IGNORE -->
```
Comment (MattHolmes123) [inline web/templates/pdf/import/cover-letter.html:30]: I wonder if this should be configurable somehow?
```
{% for line in ilb_address %}
  <p class="address-line">{{line}}</p>
{% endfor %}
```
Could always sort it later I suppose.

File: web/templates/pdf/import/cover-letter.html [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        </div>
+        <div class="header-block" id="dbt-url">
+          <span class="url">gov.uk/dbt</span>
```
Comment (MattHolmes123) [inline web/templates/pdf/import/cover-letter.html:33]: I get page not found going to this url: `https://www.gov.uk/dbt`

File: web/templates/pdf/import/cover-letter.html [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      <div class="endorsements-block page-margin">
+        {% for endorsement in process.endorsements.all() %}
+          <p>{{ endorsement.content }}</p>
+        {% endfor %}
+      </div>
+    {% elif preview %}
+      <div class="endorsements-block page-margin">
+        <p class="additional-remarks">[[ ADDITIONAL_REMARKS ]]</p>
+      </div>
+    {% endif %}
```
Comment (MattHolmes123) [inline web/templates/pdf/import/cover-letter.html:59]: If this were me I wouldn't have this `if else endif` logic in the template.
I'd supply a list of endorsements in the context.
e.g. something like this
```
if preview:
    endorsements = ["[[ ADDITIONAL_REMARKS ]]"]
else:
    endorsements = process.endorsements.values_list("content", flat=True")
```
The less logic in the template the better - templates should be as dumb as they can be

File: web/tests/domains/template/test_utils.py [hunk 7]
```diff
@@ -3,26 +3,49 @@
 
 import pytest
 
+from web.domains.case._import.fa_dfl.models import DFLApplication
```
Comment (MattHolmes123) [inline web/tests/domains/template/test_utils.py:6]: we can import all models from `web.models` .

Comment (MattHolmes123) [reply web/tests/domains/template/test_utils.py:6]: Just seen further down you've deleted the existing imports from web.models I don't understand why.
Importing from web.models allows for easier refactoring as if you change the underlying structure web.models isn't going anywhere

**Review by MattHolmes123 — CHANGES_REQUESTED** [review]

File: web/domains/template/context.py [hunk 8]
```diff
@@ -1,12 +1,24 @@
...
+
+    def __getitem__(self, item: str) -> str:
+        ...
+
 
 class CoverLetterTemplateContext:
     date_fmt = "%d %B %Y"
 
-    def __init__(self, application: ImportApplication, document_type: DocumentTypes):
+    def __init__(self, application: "ImportApplication", document_type):
```
[the remark this answers is not in the document]
Comment (mikeylockwood) [reply web/domains/template/context.py:21]: Trying to figure out why DocumentTypes was causing a circular import and forgot to put it back in

File: web/tests/domains/template/test_utils.py (hunk 7 above)
Comment (mikeylockwood) [reply web/tests/domains/template/test_utils.py:6]: It was again part of trying to figure out the circular import

File: web/templates/pdf/import/cover-letter.html (hunk 1 above)
Comment (mikeylockwood) [reply web/templates/pdf/import/cover-letter.html:4]: Yes as it doesn't inherit from the base template. The favicon image seems to just be a placeholder at the minute so will need to be replaced at some point

File: web/templates/pdf/import/cover-letter.html (hunk 6 above)
Comment (mikeylockwood) [reply web/templates/pdf/import/cover-letter.html:59]: I want different classes for the p elements still. It is nicer just to do this in the template

File: web/templates/pdf/import/cover-letter.html (hunk 5 above)
Comment (mikeylockwood) [reply web/templates/pdf/import/cover-letter.html]: I'll just update wit when it's created if that makes you happy. The dit link says this page has a big banner saying this organisation is changing so will need to be updated

File: web/templates/pdf/import/cover-letter.html (hunk 6 above)
Comment (MattHolmes123) [reply web/templates/pdf/import/cover-letter.html:54]: Oh yer I just noticed the extra class on the p tag.
You could still do that without the neeed for the if else and django db logic in the template.

**Review by MattHolmes123 — APPROVED** [review]

**Review by MattHolmes123 — COMMENTED** [review]
It is generating cover letters for sanctions applications

**Pull request merged**